Detectors/TPC/reconstruction/src/TPCFastTransformManagerO2.cxx

```diff
@@ -0,0 +1,261 @@
+      knotsV[4+i] = d1 + i/9. * d2;
+    }
+    for( int i=0; i<5; i++ ){ // 5 bins for last 90% <-> 100%
+      knotsV[13+i] = d1 + d2 + i/4. * d3;
+    }
+
+    spline.construct( nKnotsU, knotsU, nAxisTicksU,
+		      nKnotsV, knotsV, nAxisTicksV );
+  }
+  distortion.setApproximationScenario( 0, spline );  
+  distortion.finishConstruction();
+  fastTransform.finishConstruction();
+
+  // check if calculated pad geometry is consistent with the map
+  testGeometry(fastTransform);
+
+  return updateCalibration( fastTransform, TimeStamp );
+}
+
+  
+int TPCFastTransformManagerO2::updateCalibration( ali_tpc_common::tpc_fast_transformation::TPCFastTransform &fastTransform, Long_t TimeStamp )
+{
+  // Update the calibration with the new time stamp
+
+  Long_t lastTS = fastTransform.getTimeStamp();
+  
+  // deinitialize
+
+  fastTransform.setTimeStamp( -1 ); 
+
+  if( TimeStamp < 0  ) return 0; 
+
+  // search for the calibration database ...
+
+  const static ParameterDetector& detParam = ParameterDetector::defaultInstance();
+  const static ParameterGas& gasParam = ParameterGas::defaultInstance();
+  const static ParameterElectronics& elParam = ParameterElectronics::defaultInstance();
+
+  // calibration found, set the initialized status back
+
+  fastTransform.setTimeStamp(lastTS);
+
+  // less than 60 seconds from the previois time stamp, don't do anything
+
+  if( lastTS>=0 && TMath::Abs(lastTS - TimeStamp ) <60 ) return 0; 
+  
+  // start the initialization
+  
+  fastTransform.setTimeStamp(TimeStamp);
+  
+  // find last calibrated time bin
+  
+  double vDrift = elParam.getZBinWidth() * gasParam.getVdrift();
+  mLastTimeBin = detParam.getTPClength() / vDrift  + 1;
+
+ 
+  // fast transform formula:
+  // L = (t-t0)*(mVdrift + mVdriftCorrY*yLab ) + mLdriftCorr 
+  // Z = Z(L) +  tpcAlignmentZ
+  // spline distortions for xyz
+  // Time-of-flight correction: ldrift += dist-to-vtx*tofCorr
+
+  double t0 = 0.;
+  double vdCorrY = 0.;
+  double ldCorr = 0.;
+  double tpcAlignmentZ = 0.;
+
+  double tofCorr = 0.;
+  double primVtxZ = 0.;
+
+  fastTransform.setCalibration( TimeStamp, t0, vDrift, vdCorrY, ldCorr, tofCorr, primVtxZ, tpcAlignmentZ);
+
+  // now calculate distortion map: dx,du,dv = ( origTransform() -> x,u,v) - fastTransformNominal:x,u,v
+
+  TPCDistortionIRS& distortion = fastTransform.getDistortionNonConst();
+  
+  // switch TOF correction off for a while
+ 
+  for( int slice=0; slice<distortion.getNumberOfSlices(); slice++){      
+    for( int row=0; row<distortion.getNumberOfRows(); row++ ){
+      //const TPCFastTransform::RowInfo &rowInfo = fastTransform.getRowInfo( row );
+      const IrregularSpline2D3D& spline = distortion.getSpline( slice, row );
+      float *data = distortion.getSplineDataNonConst(slice,row);
+      for( int knot=0; knot<spline.getNumberOfKnots(); knot++ ){
+	data[3*knot+0] = 0.f;
+	data[3*knot+1] = 0.f;
+	data[3*knot+2] = 0.f;	
+      } // knots      
+      spline.correctEdges(data);
```

Comment:
Yes, _data_ is changed. spline corrects its part of the _data_ array. 
There will be some initialisation here which is specific for the _spline_. The _correctEdges()_ is a part of this _spline_-specific initialization.  
I can not hide this initialisation in _distortion_ class, therefore I would also not hide the _correctEdges()_ there to avoid confusion. This _correctEdges()_ should be called only once.